### ISSUE TITLE
Implement kernel stacking for OpenCL backend

### ISSUE THREAD

**Issue by opti-mix** [pytorch/glow#2888]
Launching a lot of small kernels is pretty expensive on GPUs.

Therefore fusing multiple kernels together (e.g. fusing simple elementwise kenels) can result in a significantly faster execution as less kernels need to be launched.

**Comment by nickgg** [comment]
Unsure of the interaction of kernel with the auto instrumentation (tracing). Not a blocker but please run some instrumented runs and describe how stacked kernels appear in the trace.

**Comment by opti-mix** [comment]
> Unsure of the interaction of kernel with the auto instrumentation (tracing). Not a blocker but please run some instrumented runs and describe how stacked kernels appear in the trace.

Good point. It won't be possible to report dedicated stats on a per-kernel basis for the original kernels which were stacked together. The whole stacked kernel would appear as a single fused kernel in the trace, I assume. Of course, its name or description may mention the original kernels it consists of, so that one at least gets an idea about the operations inside this kernel.

**Comment by nickgg** [comment]
@ZchiPitt's work on DebugInfo should be useful here since this is an example of a backend specific transformation.
